paasta_tools/setup_prometheus_adapter_config.py

```diff
@@ -228,6 +231,119 @@ def should_create_piscina_scaling_rule(
+        "desired_active_requests_per_replica"
+    ]
+    moving_average_window = autoscaling_config.get(
+        "moving_average_window_seconds",
+        DEFAULT_ACTIVE_REQUESTS_AUTOSCALING_MOVING_AVERAGE_WINDOW,
+    )
+    deployment_name = get_kubernetes_app_name(service=service, instance=instance)
+
+    # In order for autoscaling to work safely while a service migrates from one namespace to another, the HPA needs to
+    # make sure that the deployment in the new namespace is scaled up enough to handle _all_ the load.
+    # This is because once the new deployment is 100% healthy, cleanup_kubernetes_job will delete the deployment out of
+    # the old namespace all at once, suddenly putting all the load onto the deployment in the new namespace.
+    # To ensure this, we must:
+    #  - DO NOT filter on namespace in worker_filter_terms (which is used when calculating desired_instances).
+    #  - DO filter on namespace in replica_filter_terms (which is used to calculate current_replicas).
+    # This makes sure that desired_instances includes load from all namespaces, but that the scaling ratio calculated
+    # by (desired_instances / current_replicas) is meaningful for each namespace.
+    worker_filter_terms = f"paasta_cluster='{paasta_cluster}',service_name='{service}',service_instance='{instance}'"
+    replica_filter_terms = f"paasta_cluster='{paasta_cluster}',deployment='{deployment_name}',namespace='{namespace}'"
+
+    current_replicas = f"""
+        sum(
+            label_join(
+                (
+                    kube_deployment_spec_replicas{{{replica_filter_terms}}} >= 0
+                    or
+                    max_over_time(
+                        kube_deployment_spec_replicas{{{replica_filter_terms}}}[{DEFAULT_EXTRAPOLATION_TIME}s]
+                    )
+                ),
+                "kube_deployment", "", "deployment"
+            )
+        ) by (kube_deployment)
+    """
+    # k8s:deployment:pods_status_ready is a metric created by summing kube_pod_status_ready
+    # over paasta service/instance/cluster. it counts the number of ready pods in a paasta
+    # deployment.
+    ready_pods = f"""
+        (sum(
+            k8s:deployment:pods_status_ready{{{worker_filter_terms}}} >= 0
+            or
+            max_over_time(
+                k8s:deployment:pods_status_ready{{{worker_filter_terms}}}[{DEFAULT_EXTRAPOLATION_TIME}s]
+            )
+        ) by (kube_deployment))
+    """
+    load_per_instance = f"""
+        avg(
+            service_instance:envoy_cluster__egress_cluster_upstream_rq_active{{{worker_filter_terms}}}
+        ) by (kube_pod, kube_deployment)
+    """
+    missing_instances = f"""
+        clamp_min(
+            {ready_pods} - count({load_per_instance}) by (kube_deployment),
+            0
+        )
+    """
+    total_load = f"""
+    (
+        sum(
+            {load_per_instance}
+        ) by (kube_deployment)
+        +
+        {missing_instances}
```

Comment:
I'm not sure if we should add a TODO here to do something better, remove the missing instances addition, or figure out if we can do better now.

(the intent behind `missing_instances` is that every instance that's not contributing metrics is assumed to be so overloaded that it can't even report metics - i.e., at 100% utilization)

Reply:
I see. In that case, we should probably do a `missing_instances * desired_active_requests_per_replica` to account for maximum load.